src/feat/feature-common.h

```diff
@@ -0,0 +1,159 @@
+  /// Returns a reference to the frame-extraction options class, which
+  /// will be part of our own options class.
+  const FrameExtractionOptions &GetFrameOptions() const {
+    return opts_.frame_opts;
+  }
+
+  /// Returns the feature dimension
+  int32 Dim();
+
+  /// Returns true if this function may inspect the raw log-energy of the signal
+  /// (before windowing and pre-emphasis); it's safe to always return true, but
+  /// setting it to false enables an optimization.
+  bool NeedRawLogEnergy() { return true; }
+
+  /// constructor from options class; it should not store a reference or pointer
+  /// to the options class but should copy it.
+  explicit ExampleFeatureComputer(const ExampleFeatureComputerOptions &opts):
+      opts_(opts) { }
+
+  /// Copy constructor; all of these classes must have one.
+  ExampleFeatureComputer(const ExampleFeatureComputer &other);
+
+  /**
+     Function that computes one frame of features from
+     one frame of signal.
+
+     @param [in] signal_raw_log_energy The log-energy of the frame of the signal
+         prior to windowing and pre-emphasis, or
+         log(numeric_limits<float>::min()), whichever is greater.  Must be
+         ignored by this function if this class returns false from
+         this->NeedRawLogEnergy().
+     @param [in] vtln_warp  The VTLN warping factor that the user wants
+         to be applied when computing features for this utterance.  Will
+         normally be 1.0, meaning no warping is to be done.  The value will
+         be ignored for feature types that don't support VLTN, such as
+         spectrogram features.
+     @param [in] signal_frame  One frame of the signal,
+       as extracted using the function ExtractWindow() using the options
+       returned by this->GetFrameOptions().  The function will use the
+       vector as a workspace, which is why it's a non-const pointer.
+     @param [out] feature  Pointer to a vector of size this->Dim(), to which
+         the computed feature will be written.
+  */
+  void Compute(BaseFloat signal_log_energy,
+               BaseFloat vtln_warp,
+               VectorBase<BaseFloat> *signal_frame,
+               VectorBase<BaseFloat> *feature);
+
+ private:
+  // disallow assignment.
+  ExampleFeatureComputer &operator = (const ExampleFeatureComputer &in);
+  Options opts_;
+};
+
+
+/// This templated class is intended for offline feature extraction, i.e. where
+/// you have access to the entire signal at the start.  It exists mainly to be
+/// drop-in replacement for the old (pre-2016) classes Mfcc, Plp and so on, for
+/// use in the offline case.  In April 2016 we reorganized the online
+/// feature-computation code for greater modularity and to have correct support
+/// for the snip-edges=false option.
+template <class F>
+class FeatureTpl {
+ public:
+  typedef typename F::Options Options;
+
+  FeatureTpl(const Options &opts):
+      computer_(opts),
+      feature_window_function_(computer_.GetFrameOptions()) { }
```

Comment:
not sure what you mean by this.

On Mon, Apr 11, 2016 at 12:01 AM, Kirill Katsnelson <
notifications@github.com> wrote:

> In src/feat/feature-common.h
> <https://github.com/kaldi-asr/kaldi/pull/679#discussion_r59161125>:
>
> > +
> > +
> > +/// This templated class is intended for offline feature extraction, i.e. where
> > +/// you have access to the entire signal at the start.  It exists mainly to be
> > +/// drop-in replacement for the old (pre-2016) classes Mfcc, Plp and so on, for
> > +/// use in the offline case.  In April 2016 we reorganized the online
> > +/// feature-computation code for greater modularity and to have correct support
> > +/// for the snip-edges=false option.
> > +template <class F>
> > +class FeatureTpl {
> > + public:
> > +  typedef typename F::Options Options;
> > +
> > +  FeatureTpl(const Options &opts):
> > +      computer_(opts),
> > +      feature_window_function_(computer_.GetFrameOptions()) { }
>
> Can one of these be named matching the other for clarity?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/679/files/8e7c759093ca984eeab029d0c6c1393249f1a158#r59161125>
>